rfcs/testdriver_bidi.md

```diff
@@ -0,0 +1,277 @@
+# RFC 185: Add WebDriver BiDi support to testdriver.js
+
+## Summary
+
+Add “testdriver.js” support for [WebDriver BiDi](https://w3c.github.io/webdriver-bidi/) events and actions. To support WebDriver BiDi, the following parts are required:
+* **Wptrunner support for asynchronous actions.** WebDriver BiDi commands are asynchronous. To implement them, the test runner must support asynchronous actions. It should support concurrent actions as well. This means all async actions can be concurrent.
+* **Wptrunner - testdriver transport support for events.** WebDriver BiDi events can be sent from the browser to the client. In order to support WebDriver BiDi events, wptrunner should be able to send events to testdriver.
+* **Extend testdriver API.** Testdriver should be extended to support WebDriver BiDi commands and events.
+* **Update RFC policy.** In [RFC policy](https://github.com/web-platform-tests/rfcs/blob/master/README.md) do not require RFC for changes extending “testdriver.js” with a method that closely matches a [WebDriver BiDi](https://w3c.github.io/webdriver-bidi) protocol.
+
+## Background
+
+The “testdriver.js” framework (or testdriver for simplicity) is based on the one-directional WebDriver Classic protocol and lacks the ability to receive events from the browser via wptrunner, limiting our ability to test event-based functionalities such as console events in WPT (e.g. many [console log tests](https://github.com/web-platform-tests/wpt/tree/master/console) are [manual](https://github.com/web-platform-tests/wpt/blob/master/console/console-timing-logging-manual.html)!). To overcome this limitation, this RFC proposes using the [WebDriver BiDi](https://w3c.github.io/webdriver-bidi) protocol, an extension for WebDriver Classic offering bidirectional communication between the test and the browser.
+
+This RFC presents changes in testdriver and wptrunner so that it incorporates both WebDriver Classic and WebDriver BiDi functionalities, demonstrating the addition of tests for console events as an [example](https://github.com/web-platform-tests/wpt/pull/44649).
+
+## Details
+
+### Glossary
+
+* **testdriver**: “testdriver.js” framework used by tests and providing a way to communicate with the browser via wptrunner.
+* **wptrunner**: backend python scripts implementing test harness, providing communication between testdriver and browser.
+* **test harness**: communication channel between testdriver and wptrunner.
+
+### Example
+
+Here is an [example](https://github.com/sadym-chromium/wpt/blob/sadym/testdriver-bidi/console/console-log-logged.html) of a test using the proposed `test_driver.bidi` API for testing console log events.
+```javascript
+promise_test(async (test) => {
+    const some_message = "SOME MESSAGE";
+    // Subscribe to `log.entryAdded` BiDi events. This will not add a listener to the page.
+    await test_driver.bidi.log.entry_added.subscribe();
+    test.add_cleanup(async function() {
```

Comment:
In order to address the concern, we need to add a dependency between testharness and testdriver. I see the following ways:
1. Make testharness aware of the events concept and handle subscriptions there. The events are still implemented in testdriver:
```
...
promise_test(async (test) => {
    ...
    const log_entry_promise = test.events["log.entryAdded"].once();
    ...
}, "Assert log event is logged", {events: {"log.entryAdded": test_driver.bidi.log.entry_added}});
```
The `test_driver` is required to allow `testharness` managing the subscriptions. The user is supposed to use events via testharness (` test.events...`). The specific API can be discussed.
2. Make testdriver aware of testharness:
```
promise_test(async (test) => {
    ...
    await test_driver.bidi.log.entry_added.subscribe(test);
    ...
}, "Assert log event is logged");
```
This will not completely address the concern of testharness relying on events.
3. Do not unsubscribe from the events.

I find the option **1** is the preferable. WDYT?

Reply:
I didn't really mean each `test()` in a page, but between top-level test files.

The problem is that if the test file is allowed unfettered access to BiDi event subscriptions we can't isolate it from any subscriptions that the harness itself depends on, and doing something like closing the bidi session after each test would be very high overhead. The most obvious example would be if we wanted to use `script.message` to drive the harness, but also allow scripts themselves to emit custom messages. But it could also happen that, for example, we wanted the harness to be able to subscribe to `console.log` messages and re-emit them in the harness logs. Allowing the test to unsubscribe from those events would break that. So probably the harness has to manage subscriptions and know which events it subscribed to, and which should be forwarded back to the browser.

Reply:
Would it be sufficient if we implement automatic clean up of events similar to how BiDi's own test suite does it https://github.com/web-platform-tests/wpt/blob/d7b9fab5017587079985d5ab29781fcf3043533e/webdriver/tests/support/fixtures_bidi.py#L50 ? we could have executewebdriver.py record all subscribe calls coming from the test file and run corresponding unsubscribe calls filtering out events required by the testharness? In that case, unsubscribe calls would not be exposed to testdriver. There will probably be some edge cases around unsubscribing context-specific and global subscriptions at the same time.

Alternative ideas would require changing the spec:

1) support multiple light-wight BiDi sessions that do not require starting/shutting down the browser.
2) returning a subscriptionId in subscribe calls so that only specific subscriptions could be cleaned up.

Reply:
I believe the API in **option 1** partly addresses the concerns you mentioned. It allows implementing event subscription and filtering in the [testharness](https://github.com/web-platform-tests/wpt/blob/91e2e520993bdc9820a0ddd619d27242f3abba88/resources/testharness.js#L1027) without exposing that logic to the test file, but the specific filtering of let's say `script.message` intended to be send to the test or to the testharness can be tricky.

> So probably the harness has to manage subscriptions and know which events it subscribed to, and which should be forwarded back to the browser.

Do you mean the backend testharness, or the client-side?

Reply:
FWIW @OrKoN's option 2 is something we'd like to propose for the BiDi spec (although it might have to be opt-in to _require_ a handle to unsubscribe since that's not how things currently work).

Reply:
The proper long-term solution would be to provide separate sessions for the testharness and for each test files. This would provide proper isolations and tear down.

As an intermediate solution, in the new revision of the [example PR](https://github.com/web-platform-tests/wpt/pull/44649/files#diff-e5a8911dd97e0352b1b26d8ce6ef0a92b25378f7c9e79371a1eb1b1834bc9a8dR128) I implemented clean up on the `wptrunner` side as a part of `BidiEventsProtocolPart`. This allowed to remove `unsubscribe` from the testdriver API. We can always add it, if we have a valid use case.

@jgraham WDYT?

Reply:
I updated the RFC and the example with the new version. The protocol part `WebDriverBidiEventsProtocolPart` has a method `unsubscribe_all`, which is responsible for keeping track of all the subscriptions.